ci: intermittent failure in Windows CI

https://github.com/bitcoin/bitcoin/actions/runs/23945053179/job/69839333752?pr=34978#step:8:587:
```bash
Downloading https://github.com/boostorg/mpl/archive/boost-1.88.0.tar.gz -> boostorg-mpl-boost-1.88.0.tar.gz
error: https://github.com/boostorg/mpl/archive/boost-1.88.0.tar.gz: failed: status code 502
note: If you are using a proxy, please ensure your proxy settings are correct.
Possible causes are:
1. You are actually using an HTTP proxy, but setting HTTPS_PROXY variable to `https://address:port`.
This is not correct, because `https://` prefix claims the proxy is an HTTPS proxy, while your proxy (v2ray, shadowsocksr, etc...) is an HTTP proxy.
Try setting `http://address:port` to both HTTP_PROXY and HTTPS_PROXY instead.
2. If you are using Windows, vcpkg will automatically use your Windows IE Proxy Settings set by your proxy software. See: https://github.com/microsoft/vcpkg-tool/pull/77
The value set by your proxy might be wrong, or have same `https://` prefix issue.
3. Your proxy's remote server is out of service.
If you believe this is not a temporary download server failure and vcpkg needs to be changed to download this file from a different location, please submit an issue to https://github.com/Microsoft/vcpkg/issues
CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:136 (message):
  Download failed, halting portfile.
Call Stack (most recent call first):
  scripts/cmake/vcpkg_from_github.cmake:120 (vcpkg_download_distfile)
  C:/Users/runneradmin/AppData/Local/vcpkg/registries/git-trees/6cdd95078ed4dc9bd700884cfc132dca24a99013/portfile.cmake:3 (vcpkg_from_github)
  scripts/ports.cmake:206 (include)

```

We have recently added logic, to try and prevent these kinds of intermittent failures, by retrying. Seen here: https://github.com/bitcoin/bitcoin/actions/runs/23945053179/job/69839333752?pr=34978#step:8:738)

Maybe we should retry more aggressively?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -267,7 +267,7 @@ jobs:
         id: vcpkg-binary-cache
         with:
           path: ~/AppData/Local/vcpkg/archives
-          key: ${{ github.job }}-vcpkg-binary-${{ hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json') }}
+          key: ${{ github.job }}-vcpkg-binary-release-${{ hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json') }}
 
       - name: Generate build system
         run: |

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -253,13 +253,14 @@ jobs:
           py -3 --version
           bash --version
 
-      - name: Restore vcpkg tools cache
-        id: vcpkg-tools-cache
+      - name: Restore vcpkg downloads cache
+        id: vcpkg-downloads-cache
         uses: actions/cache/restore@v5
         with:
-          path: ~/AppData/Local/vcpkg/downloads/tools
-          key: ${{ github.job }}-vcpkg-tools-${{ github.run_id }}
-          restore-keys: ${{ github.job }}-vcpkg-tools-
+          path: |
+            ~/AppData/Local/vcpkg/downloads/*
+            !~/AppData/Local/vcpkg/downloads/tools
+          key: ${{ github.job }}-vcpkg-downloads-${{ hashFiles('vcpkg.json') }}
 
       - name: Restore vcpkg binary cache
         uses: actions/cache/restore@v5
@@ -279,13 +280,15 @@ jobs:
           path: ~/AppData/Local/vcpkg/archives
           key: ${{ steps.vcpkg-binary-cache.outputs.cache-primary-key }}
 
-      - name: Save vcpkg tools cache
+      - name: Save vcpkg downloads cache
         uses: actions/cache/save@v5
-        # Only save cache from one job as they share tools. If the matrix is expanded to jobs with unique tools, this may need amending.
-        if: github.event_name != 'pull_request' && github.ref_name == github.event.repository.default_branch && steps.vcpkg-tools-cache.outputs.cache-hit != 'true' && matrix.job-type == 'standard'
+        # Only save cache from the 'standard' job, as it includes the necessary downloads for other jobs in the matrix. If the matrix is modified, this may need amending.
+        if: github.event_name != 'pull_request' && github.ref_name == github.event.repository.default_branch && steps.vcpkg-binary-cache.outputs.cache-hit != 'true' && steps.vcpkg-downloads-cache.outputs.cache-hit != 'true' && matrix.job-type == 'standard'
         with:
-          path: ~/AppData/Local/vcpkg/downloads/tools
-          key: ${{ steps.vcpkg-tools-cache.outputs.cache-primary-key }}
+          path: |
+            ~/AppData/Local/vcpkg/downloads/*
+            !~/AppData/Local/vcpkg/downloads/tools  # Cache the tools once as archives, but not redundantly in extracted form.
+          key: ${{ steps.vcpkg-downloads-cache.outputs.cache-primary-key }}
 
       - name: Build
         run: |

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -211,7 +211,7 @@ jobs:
         with:
           path: ${{ env.CCACHE_DIR }}
           # https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache
-          key: ${{ github.job }}-${{ matrix.job-type }}-ccache-${{ github.run_id }}
+          key: ${{ steps.ccache-cache.outputs.cache-primary-key }}
 
   windows-native-dll:
     name: ${{ matrix.job-name }}
@@ -277,7 +277,7 @@ jobs:
         if: github.event_name != 'pull_request' && github.ref_name == github.event.repository.default_branch && steps.vcpkg-binary-cache.outputs.cache-hit != 'true' && matrix.job-type == 'standard'
         with:
           path: ~/AppData/Local/vcpkg/archives
-          key: ${{ github.job }}-vcpkg-binary-${{ hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json') }}
+          key: ${{ steps.vcpkg-binary-cache.outputs.cache-primary-key }}
 
       - name: Save vcpkg tools cache
         uses: actions/cache/save@v5
@@ -285,7 +285,7 @@ jobs:
         if: github.event_name != 'pull_request' && github.ref_name == github.event.repository.default_branch && steps.vcpkg-tools-cache.outputs.cache-hit != 'true' && matrix.job-type == 'standard'
         with:
           path: ~/AppData/Local/vcpkg/downloads/tools
-          key: ${{ github.job }}-vcpkg-tools-${{ github.run_id }}
+          key: ${{ steps.vcpkg-tools-cache.outputs.cache-primary-key }}
 
       - name: Build
         run: |

